[move-prover] support old(..) in inlined spec blocks

mengxu-fb opened the pull request:
In addition to post states, old(..) is now allowed in `assert` and
`assume` statements in inlined spec blocks. The primary use case is to
support loop invariants: as showcased in the `verify_vector.move` test.

Note that the old(..) expressions in inline spec block is restricted to
take a single expression that is a function parameter only. No other
cases are currently allowed. Should we want to relax this constraint in
the future, the `ModuleBuilder::check_condition_is_valid` function needs
to be updated and the test case `inline_spec_err.move` in `move-model`
needs to be updated as well.

## Motivation

Complete the loop invariant story.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/main/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

- CI
- `inline_spec_err.move` is added to test that only a single function parameter expression is allowed in `old(..)` in inline spec
- `specs_in_fun_ref.move` is updated to test that we can reason about local mut refs.
- `verify_vector.move` is updated to use the `old(..)` expressions to complete the specification and verification of loop invariants. The default flow verifies successfully but the cvc4 version failed. Please refer to the `.cvc4_exp` file for details.

## Related PRs

(I think the spec-lang doc should be updated as well, I'll link that once updated)

mengxu-fb commented on language/move-model/src/spec_translator.rs:
```diff
@@ -577,7 +598,7 @@ impl<'a, 'b, T: ExpGenerator<'a>> ExpRewriterFunctions for SpecTranslator<'a, 'b
     fn rewrite_temporary(&mut self, id: NodeId, idx: TempIndex) -> Option<Exp> {
         // Compute the effective index.
-        let is_mut = self.fun_env.get_local_type(idx).is_mutable_reference();
+        let is_mut = self.builder.get_local_type(idx).is_mutable_reference();
```
This is a bug fix. As the `builder` may have more temporaries than what we recorded in `fun_env`.

mengxu-fb replied:
Well, this is probably not a bug, using `fun_env` and `builder` gives different types for the same local idx in a way I did not fully understand... The unit test failure is caused by this and I am still checking.

mengxu-fb replied:
Last update:
- use `fun_env` if `idx < param_count`: we want to be aware whether the param is a mut ref or not.
- use `builder` otherwise.

wrwg reviewed (dismissed):
Would you mind checking if we already have a test for a local &mut used from a code spec block, and if not, add one?

mengxu-fb commented:
@wrwg I updated the PR in two ways:

- Restricted `old(..)` expressions in inline spec block to take a single expression that is a function parameter only. Hopefully this might clear some confusion on what does `old(..)` mean when applied to a local variable. We can gradually relax this constraint when we find other use cases for `old(..)` in inline specs.

- Added two test cases, one for the above restriction on `old(..)` expressions. The other for testing that we can prove properties on local &mut and `old(..)`.

mengxu-fb commented:
/land
